File: PianoKeyboard_java_eclipse/src/com/teuskim/pianokeyboard/RecommendWord.java
package com.teuskim.pianokeyboard;

public class RecommendWord implements Comparable<RecommendWord> {
	
	public int mWordId;
	public String mWord;
	public double mUseCntNext;
	public double mUseCntNextSum;
	public double mUseCntXxx;
	public double mUseCntXxxSum;
	public double mUseCntTotal;
	public double mUseCntTotalSum;
	public double mUseCntN;
	public double mUseCntNSum;
	
	public double mPoint;
	
	public String mUpdDt;
	public long mDiffTime;
	
	public void generatePoint(double affinityWeight, double useCntXxxWeight, double useCntTotalWeight, double useCntNWeight, long availableTime){
		double p1 = (mUseCntNextSum > 0) ? (mUseCntNext / mUseCntNextSum) : 0;
		double p2 = (mUseCntXxxSum > 0) ? (mUseCntXxx / mUseCntXxxSum) : 0;
		double p3 = (mUseCntTotalSum > 0) ? (mUseCntTotal / mUseCntTotalSum) : 0;
		double p4 = (mUseCntNSum > 0) ? (mUseCntN / mUseCntNSum) : 0;
		double w1 = affinityWeight;
		double w2 = useCntXxxWeight;
		double w3 = useCntTotalWeight;
		double w4 = useCntNWeight;
//		Log.e("AAAA", "generatePoint: "+mWord+" , "+p1+","+p2+","+p3+","+p4+" -- "+w1+","+w2+","+w3+","+w4);
		
		mPoint = (p1*w1) + (p2*w2) + (p3*w3) + (p4*w4);
		if(mDiffTime > 0 && mDiffTime > availableTime)
			mPoint = mPoint * 0.5;
	}
	
	public void generateDiffTime(long currTime){
		if(mUpdDt != null && mUpdDt.length() > 0){
			mDiffTime = currTime - Long.valueOf(mUpdDt);
		}
	}
	
	@Override
	public int compareTo(RecommendWord another) {
		double diff = mPoint - another.mPoint;
		int result = 0;
		if(diff < 0)
			result = 1;
		else if(diff > 0)
			result = -1;
		return result;
	}
	
}

File: PianoKeyboard_kotlin/app/src/main/java/com/teuskim/pianokeyboard/RecommendWord.kt
package com.teuskim.pianokeyboard

class RecommendWord : Comparable<RecommendWord> {

    var mWordId: Int = 0
    var mWord: String? = null
    var mUseCntNext: Double = 0.0
    var mUseCntNextSum: Double = 0.0
    var mUseCntXxx: Double = 0.0
    var mUseCntXxxSum: Double = 0.0
    var mUseCntTotal: Double = 0.0
    var mUseCntTotalSum: Double = 0.0
    var mUseCntN: Double = 0.0
    var mUseCntNSum: Double = 0.0

    var mPoint: Double = 0.0

    var mUpdDt: String? = null
    var mDiffTime: Long = 0

    fun generatePoint(affinityWeight: Double, useCntXxxWeight: Double, useCntTotalWeight: Double, useCntNWeight: Double, availableTime: Long) {
        val p1: Double = if (mUseCntNextSum > 0) mUseCntNext / mUseCntNextSum else 0.0
        val p2: Double = if (mUseCntXxxSum > 0) mUseCntXxx / mUseCntXxxSum else 0.0
        val p3: Double = if (mUseCntTotalSum > 0) mUseCntTotal / mUseCntTotalSum else 0.0
        val p4: Double = if (mUseCntNSum > 0) mUseCntN / mUseCntNSum else 0.0
//		Log.e("AAAA", "generatePoint: "+mWord+" , "+p1+","+p2+","+p3+","+p4+" -- "+w1+","+w2+","+w3+","+w4);

        mPoint = p1 * affinityWeight + p2 * useCntXxxWeight + p3 * useCntTotalWeight + p4 * useCntNWeight
        if (mDiffTime > 0 && mDiffTime > availableTime)
            mPoint = mPoint * 0.5
    }

    fun generateDiffTime(currTime: Long) {
        if (mUpdDt != null && mUpdDt!!.length > 0) {
            mDiffTime = currTime - java.lang.Long.valueOf(mUpdDt)!!
        }
    }

    override fun compareTo(another: RecommendWord): Int {
        val diff = mPoint - another.mPoint
        var result = 0
        if (diff < 0)
            result = 1
        else if (diff > 0)
            result = -1
        return result
    }

}
